### PR TITLE
Don't use exceptions in low-level mem allocator

### PR DESCRIPTION
**Pull request by hyc**
Too annoying in log messages for non-exceptional conditions.

### CODE REVIEW

**Comment by hyc** [comment]
Using C for `virtual_memory` will also remove the need for the linker/nmedit trick in #240.

**Comment by hyc** [comment]
... actually the relevant part of #240 is already included here, so we can close that one.

**Review by SChernykh — COMMENTED** [review]
I see `errfunc` filled in but not returned to the outside code in any way, what is the point of having it?

File: src/virtual_memory.c [hunk 1]
```diff
@@ -53,27 +53,43 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 	BOOL             status;
 	DWORD            error;
 
 	if (!OpenProcessToken(GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES | TOKEN_QUERY, &hToken))
-		throw std::runtime_error(getErrorMessage("OpenProcessToken"));
+		Fail("OpenProcessToken");
 
 	if (!LookupPrivilegeValue(NULL, pszPrivilege, &tp.Privileges[0].Luid))
-		throw std::runtime_error(getErrorMessage("LookupPrivilegeValue"));
+		Fail("LookupPrivilegeValue");
```
Comment (SChernykh) [inline src/virtual_memory.c:92]: `hToken` will leak here, you need to `CloseHandle(hToken);` at every function `return`

**Comment by hyc** [comment]
> 

The error info was never propagated directly back to the caller, it was only embedded into the text of the thrown exceptions. I just tried to preserve some of the info so it can be seen in a debugger, if needed. The majority of these error cases would never happen anyway.

File: src/virtual_memory.c (hunk 1 above)
Comment (hyc) [reply src/virtual_memory.c]: thanks, fixed

**Review by SChernykh — APPROVED** [review]

File: src/virtual_memory.c [hunk 2]
```diff
@@ -53,27 +53,48 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+		MacOSchecked = 1;
+	}
+	return MacOSver >= ((major << 8) | minor);
 }
+#endif
 
-void setPrivilege(const char* pszPrivilege, BOOL bEnable) {
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+#define Fail(func)	do  {*errfunc = func; return GetLastError()} while(0)
```
[the remark this answers is not in the document]
Comment (hyc) [reply src/virtual_memory.c]: Thanks, fixed

File: src/virtual_memory.c [hunk 3]
```diff
@@ -85,20 +106,28 @@ void setPrivilege(const char* pszPrivilege, BOOL bEnable) {
...
 }
+#else
+#define Fail(func)	do  {*errfunc = func; return errno;} while(0)
 #endif
 
-void* allocMemoryPages(std::size_t bytes) {
+void* allocMemoryPages(size_t bytes) {
 	void* mem;
 #if defined(_WIN32) || defined(__CYGWIN__)
 	mem = VirtualAlloc(nullptr, bytes, MEM_COMMIT, PAGE_READWRITE);
```
[the remark this answers is not in the document]
Comment (hyc) [reply src/virtual_memory.c]: right

**Review by tevador — APPROVED** [review]

**Comment by hyc** [comment]
Squashed

**Pull request merged**